src/CredentialRepository.php

```diff
@@ -3,99 +3,213 @@
+            return [];
+        }
+
+        return array_map(function($credentialComposite) {
+            return $credentialComposite['source'];
+        }, $this->credentials);
+    }
+
+    public function saveCredentialSource(PublicKeyCredentialSource $publicKeyCredentialSource): void
+    {
+        $ref = $this->getCredentialIDRef($publicKeyCredentialSource->getPublicKeyCredentialId());
+
+        if (!isset($this->credentials[$ref])) {
+            $this->credentials[$ref] = [
+                'counter' => 0,
+            ];
         }
+
+        $this->credentials[$ref]['source'] = $publicKeyCredentialSource;
+        $this->hasChanged = true;
+    }
+
+    /**
+     * Indicates the repository has changed and should be persisted (as this doesn't follow the actual repository
+     * pattern and is expected to be stored on a dataobject for persistence)
+     *
+     * @return bool
+     */
+    public function hasChanged(): bool
+    {
+        return $this->hasChanged;
+    }
+
+    /**
+     * Set the credentials in bulk (for internal use) ensuring that credential objects are initialised correctly
+     *
+     * @param array $credentials
+     */
+    protected function setCredentials(array $credentials): void
+    {
+        $this->credentials = array_map(function ($data) {
+            $data['source'] = PublicKeyCredentialSource::createFromArray($data['source']);
+            return $data;
+        }, $credentials);
+    }
+
+    /**
+     * Create a reference to be used as a key for the credentials in the array
+     *
+     * @param string $credentialID
+     * @return string
+     */
+    protected function getCredentialIDRef(string $credentialID): string
+    {
+        return base64_encode($credentialID);
+    }
+
+    /**
+     * Provide the credentials stored in this repository as an array
+     *
+     * @return array
+     */
+    public function toArray(): array
+    {
+        return $this->credentials;
+    }
+
+    /**
+     * Create an instance of a repository from the given credentials
+     *
+     * @param array $credentials
+     * @param string $memberID
+     * @return CredentialRepository
+     */
+    public static function fromArray(array $credentials, string $memberID): self
+    {
+        $new = new static($memberID);
+        $new->setCredentials($credentials);
+
+        return $new;
+    }
+
+    /**
+     * String representation of object
+     * @link https://php.net/manual/en/serializable.serialize.php
+     * @return string the string representation of the object or null
+     * @since 5.1.0
+     */
+    public function serialize()
+    {
+        return json_encode(['credentials' => $this->toArray(), 'memberID' => $this->memberID]);
+    }
+
+    /**
+     * Constructs the object
+     * @link https://php.net/manual/en/serializable.unserialize.php
+     * @param string $serialized <p>
+     * The string representation of the object.
+     * </p>
+     * @return void
+     * @since 5.1.0
```

Comment:
These doc blocks seem to be from PHP internals - maybe remove? They should inherit from the interface if you don't define them